components/sup/src/http_gateway.rs

```diff
@@ -0,0 +1,194 @@
+
+impl Default for ListenAddr {
+    fn default() -> ListenAddr {
+        ListenAddr(SocketAddrV4::new(Ipv4Addr::new(0, 0, 0, 0), 9631))
+    }
+}
+
+impl Deref for ListenAddr {
+    type Target = SocketAddrV4;
+
+    fn deref(&self) -> &SocketAddrV4 {
+        &self.0
+    }
+}
+
+impl DerefMut for ListenAddr {
+    fn deref_mut(&mut self) -> &mut SocketAddrV4 {
+        &mut self.0
+    }
+}
+
+impl FromStr for ListenAddr {
+    type Err = SupError;
+
+    fn from_str(val: &str) -> Result<Self> {
+        match SocketAddrV4::from_str(val) {
+            Ok(addr) => Ok(ListenAddr(addr)),
+            Err(_) => {
+                match Ipv4Addr::from_str(val) {
+                    Ok(ip) => {
+                        let mut addr = ListenAddr::default();
+                        addr.set_ip(ip);
+                        Ok(addr)
+                    }
+                    Err(_) => Err(sup_error!(Error::IPFailed)),
+                }
+            }
+        }
+    }
+}
+
+struct ManagerState;
+
+impl typemap::Key for ManagerState {
+    type Value = manager::State;
+}
+
+pub struct Server(Iron<Chain>);
+
+impl Server {
+    pub fn new(manager_state: manager::State) -> Self {
+        let router = router!(
+            butterfly: get "/butterfly" => butterfly,
+            census: get "/census" => census,
+            services: get "/services" => services,
+            service_config: get "/services/:svc/:group/config" => config,
+            service_health: get "/services/:svc/:group/health" => health,
+            service_config_org: get "/services/:svc/:group/:org/config" => config,
+            service_health_org: get "/services/:svc/:group/:org/health" => health,
+        );
+        let mut chain = Chain::new(router);
+        chain.link(persistent::Read::<ManagerState>::both(manager_state));
+        Server(Iron::new(chain))
+    }
+
+    pub fn start(self) -> Result<JoinHandle<()>> {
+        let handle = try!(thread::Builder::new()
+            .name("http-gateway".to_string())
+            .spawn(move || {
+                self.0
+                    .http(*gconfig().http_listen_addr())
+                    .expect("unable to start http-gateway thread");
+            }));
+        Ok(handle)
+    }
+}
+
+fn butterfly(req: &mut Request) -> IronResult<Response> {
+    let state = req.get::<persistent::Read<ManagerState>>().unwrap();
+    Ok(Response::with((status::Ok, json::encode(&state.butterfly).unwrap())))
+}
+
+fn census(req: &mut Request) -> IronResult<Response> {
+    let state = req.get::<persistent::Read<ManagerState>>().unwrap();
+    let data = state.census_list.read().unwrap();
+    Ok(Response::with((status::Ok, json::encode(&*data).unwrap())))
+}
+
+fn config(req: &mut Request) -> IronResult<Response> {
+    let state = req.get::<persistent::Read<ManagerState>>().unwrap();
+    let service_group =
+        ServiceGroup::new(req.extensions.get::<Router>().unwrap().find("svc").unwrap(),
```

Comment:
What happens when this panics? Can it essentially not panic because of the shape of the route?